### PR TITLE
[Feature] support recover for cloud native pk table

### PR DESCRIPTION
**Pull request by luohaha**
Why I'm doing:

When tablet's inconsistent state happens, such as:
 1. Failure of primary key uniqueness constraints. Duplicate primary keys show up.
 2. Internal data correctness check failure.

These anomalies show up (Because of Bug or something), and user don't want to drop this partition, we need a way to recover from these anomalies.

What I'm doing:

This PR supports a way to recover from these situations, by `PrimaryKeyRecover`. 
 `PrimaryKeyRecover` will regard segment files in tablet meta as single source of truth, remove old delvec files and pk index., and then rebuild delvec files and new pk index from scratch.

## What type of PR is this:

- [ ] BugFix
- [x] Feature
- [ ] Enhancement
- [ ] Refactor
- [ ] UT
- [ ] Doc
- [ ] Tool

Does this PR entail a change in behavior?

- [ ] Yes, this PR will result in a change in behavior.
- [x] No, this PR will not result in a change in behavior.

If yes, please specify the type of change:

- [ ] Interface/UI changes: syntax, type conversion, expression evaluation, display information
- [ ] Parameter changes: default values, similar parameters but with different default values
- [ ] Policy changes: use new policy to replace old one, functionality automatically enabled
- [ ] Feature removed
- [ ] Miscellaneous: upgrade & downgrade compatibility, etc.

## Checklist:

- [x] I have added test cases for my bug fix or my new feature
- [ ] This pr needs user documentation (for new or modified features or behaviors)
  - [ ] I have added documentation for my new feature or new function

## Bugfix cherry-pick branch check:

- [x] I have checked the version labels which the pr will be auto-backported to the target branch
  - [x] 3.2
  - [x] 3.1
  - [ ] 3.0
  - [ ] 2.5

### CODE REVIEW

File: be/src/storage/lake/primary_key_recover.cpp [hunk 1]
```diff
@@ -0,0 +1,131 @@
...
+    rowids.reserve(DEFAULT_CHUNK_SIZE);
+    PrimaryIndex::DeletesMap new_deletes;
+    auto chunk_shared_ptr = ChunkHelper::new_chunk(_pkey_schema, DEFAULT_CHUNK_SIZE);
+    auto chunk = chunk_shared_ptr.get();
+    // 2. scan all rowsets and segments to build primary index
+    auto rowsets = _tablet->get_rowsets(*_metadata);
+    if (!rowsets.ok()) {
+        return rowsets.status();
+    }
+    // 3. rebuild primary key index and generate delete maps
```
Comment (TszKitLo40) [inline be/src/storage/lake/primary_key_recover.cpp:81]: It seems that this logic is used in many places. How about make it a function?

Comment (luohaha) [reply be/src/storage/lake/primary_key_recover.cpp:81]: Good idea, will do it.

File: be/src/storage/lake/primary_key_recover.cpp [hunk 2]
```diff
@@ -0,0 +1,131 @@
...
+    std::vector<uint32_t> rowids;
+    rowids.reserve(DEFAULT_CHUNK_SIZE);
+    PrimaryIndex::DeletesMap new_deletes;
+    auto chunk_shared_ptr = ChunkHelper::new_chunk(_pkey_schema, DEFAULT_CHUNK_SIZE);
+    auto chunk = chunk_shared_ptr.get();
+    // 2. scan all rowsets and segments to build primary index
+    auto rowsets = _tablet->get_rowsets(*_metadata);
+    if (!rowsets.ok()) {
+        return rowsets.status();
+    }
```
Comment (wyb) [inline be/src/storage/lake/primary_key_recover.cpp:80]:
```suggestion
    ASSIGN_OR_RETURN(auto rowsets, _tablet->get_rowsets(*_metadata));
```

File: be/src/storage/lake/primary_key_recover.cpp [hunk 3]
```diff
@@ -0,0 +1,131 @@
...
+            return res.status();
+        }
+        auto& itrs = res.value();
+        CHECK(itrs.size() == rowset->num_segments()) << "itrs.size != num_segments";
+        for (size_t i = 0; i < itrs.size(); i++) {
+            auto itr = itrs[i].get();
+            uint32_t row_id_start = 0;
+            if (itr == nullptr) {
+                continue;
+            }
```
Comment (wyb) [inline be/src/storage/lake/primary_key_recover.cpp:94]:
```suggestion
            auto itr = itrs[i].get();
            if (itr == nullptr) {
                continue;
            }
            uint32_t row_id_start = 0;
```

File: be/src/storage/lake/primary_key_recover.h [hunk 4]
```diff
@@ -0,0 +1,71 @@
...
+            : _builder(builder), _tablet(tablet), _metadata(metadata) {}
+    ~PrimaryKeyRecover() {}
+
+    // Follow the steps below:
+    // 1. reset_state
+    // 2. rebuild
+    //
+    // clean up delvec and primary index
+    Status pre_cleanup();
+    // delete pk index and delvec, then rebuild them
```
Comment (wyb) [inline be/src/storage/lake/primary_key_recover.h:56]:
```suggestion
    // rebuild delvec and primary index
```

File: be/src/storage/lake/primary_key_recover.cpp [hunk 5]
```diff
@@ -0,0 +1,131 @@
...
+    _metadata->clear_delvec_meta();
+    // 2. reset primary index
+    _tablet->update_mgr()->remove_primary_index_cache(_tablet->id());
+    DataDir* data_dir = StorageEngine::instance()->get_persistent_index_store(_tablet->id());
+    if (data_dir != nullptr) {
+        // clear local persistent index's index meta from rocksdb, and index files.
+        std::string tablet_pk_path =
+                strings::Substitute("$0/$1/", data_dir->get_persistent_index_path(), _tablet->id());
+        RETURN_IF_ERROR(fs::remove_all(tablet_pk_path));
+        RETURN_IF_ERROR(TabletMetaManager::remove_tablet_persistent_index_meta(data_dir, _tablet->id()));
```
[the remark this answers is not in the document]
Comment (luohaha) [reply be/src/storage/lake/primary_key_recover.cpp]: Done

**Review by wyb — DISMISSED** [review]

**Review by TszKitLo40 — APPROVED** [review]

**Review by sduzh — APPROVED** [review]

**Review by wyb — APPROVED** [review]

**Comment by luohaha** [comment]
https://github.com/Mergifyio backport branch-3.2

**Comment by luohaha** [comment]
https://github.com/Mergifyio backport branch-3.1

**Pull request merged**